packages/itwin/grouping-mapping-widget/src/widget/components/GroupPropertyUtils.ts

```diff
@@ -0,0 +1,364 @@
+      propertyTraversal: newPropertyTraversal,
+      type,
+      primitiveNavigationClass,
+      actualECClassName,
+      parentPropertyClassName,
+      key: `${actualECClassName}__${newPropertyTraversal.join("_")}`,
+      categoryLabel: propertyField.category.label,
+    }
+  );
+});
+
+const extractPrimitiveStructProperties = (
+  propertyTraversal: string[],
+  members: StructFieldMemberDescription[],
+  categoryLabel: string,
+  actualECClassName: string,
+  parentPropertyClassName?: string,
+): PropertyMetaData[] => members.flatMap((member) => {
+  if (member.type.valueFormat === PropertyValueFormat.Primitive) {
+    const propertyName = member.name;
+    const label = member.label;
+    const type = convertType(member.type.typeName);
+    const newPropertyTraversal = [...propertyTraversal, propertyName];
+
+    return ({
+      label,
+      schema: "*",
+      className: "*",
+      propertyTraversal: newPropertyTraversal,
+      type,
+      primitiveNavigationClass: "",
+      actualECClassName,
+      parentPropertyClassName,
+      key: `${actualECClassName}__${newPropertyTraversal.join("_")}`,
+      categoryLabel,
+    });
+
+  } else if (member.type.valueFormat === PropertyValueFormat.Struct) {
+    return extractPrimitiveStructProperties(
+      propertyTraversal,
+      member.type.members,
+      categoryLabel,
+      actualECClassName,
+      parentPropertyClassName
+    );
+  }
+
+  return [];
+});
+
+const extractStruct = (property: Field) => {
+  if (property.type.valueFormat !== PropertyValueFormat.Struct) {
+    return [];
+  }
+
+  const columnName = (property as PropertiesField).properties[0]
+    .property.name;
+  const actualECClassName = (property as PropertiesField).properties[0].property.classInfo.name;
+  return extractPrimitiveStructProperties(
+    [columnName],
+    property.type.members,
+    property.category.label,
+    actualECClassName
+  );
+};
+
+const extractNested = (propertyTraversal: string[], propertyFields: Field[]): PropertyMetaData[] =>
+  propertyFields.flatMap((property) => {
+    // Generate base ECProperty
+    switch (property.type.valueFormat) {
+      case PropertyValueFormat.Primitive: {
+        return extractPrimitives([...propertyTraversal], property as PropertiesField);
+      }
+      // Get structs
+      case PropertyValueFormat.Struct: {
+        const nestedContentField = property as NestedContentField;
+        // Only handling single path and not handling nested content fields within navigations
+        if (
+          nestedContentField.pathToPrimaryClass &&
+          nestedContentField.pathToPrimaryClass.length > 1
+        ) {
+          break;
+        }
+
+        switch (nestedContentField.relationshipMeaning) {
+          // Aspects are not looked at again. Extraction does not support aspects within navigations.
+          // Navigation properties
+          case RelationshipMeaning.RelatedInstance: {
+            if (
+              // Deal with a TypeDefinition
```

Comment:
Why is there a special case for TypeDefinition?

Reply:
I have not seen another type of a related instance that weren't the primitive navigations.